From 1be82a31f13b5492813cfdfb20234725a9e1e809 Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Sat, 21 Sep 2013 12:33:09 +1000
Subject: [PATCH] Allow Disabling of 1.6.3 Structure Saving


diff --git a/src/main/java/net/minecraft/server/StructureGenerator.java b/src/main/java/net/minecraft/server/StructureGenerator.java
index 93e1c0097..0a6e5b327 100644
--- a/src/main/java/net/minecraft/server/StructureGenerator.java
+++ b/src/main/java/net/minecraft/server/StructureGenerator.java
@@ -143,7 +143,13 @@ public abstract class StructureGenerator extends WorldGenBase {
 
     protected void a(World world) {
         if (this.a == null && world != null) {
-            this.a = (PersistentStructure) world.a(PersistentStructure.class, this.a());
+            // Spigot Start
+            if (world.spigotConfig.saveStructureInfo) {
+                this.a = (PersistentStructure) world.a(PersistentStructure.class, this.a());
+            } else {
+                this.a = new PersistentStructure(this.a());
+            }
+            // Spigot End
             if (this.a == null) {
                 this.a = new PersistentStructure(this.a());
                 world.a(this.a(), (PersistentBase) this.a);
diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index 07a883f19..9df9e0131 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -191,4 +191,16 @@ public class SpigotWorldConfig
         randomLightUpdates = getBoolean( "random-light-updates", false );
         log( "Random Lighting Updates: " + randomLightUpdates );
     }
+
+    public boolean saveStructureInfo;
+    private void structureInfo()
+    {
+        saveStructureInfo = getBoolean( "save-structure-info", true );
+        log( "Structure Info Saving: " + saveStructureInfo );
+        if ( !saveStructureInfo )
+        {
+            log( "*** WARNING *** You have selected to NOT save structure info. This may cause structures such as fortresses to not spawn mobs!" );
+            log( "*** WARNING *** Please use this option with caution, SpigotMC is not responsible for any issues this option may cause in the future!" );
+        }
+    }
 }
-- 
2.14.1

